src/rttr/detail/library/library_p.h

```diff
@@ -0,0 +1,155 @@
+*                                                                                   *
+*   This file is part of RTTR (Run Time Type Reflection)                            *
+*   License: MIT License                                                            *
+*                                                                                   *
+*   Permission is hereby granted, free of charge, to any person obtaining           *
+*   a copy of this software and associated documentation files (the "Software"),    *
+*   to deal in the Software without restriction, including without limitation       *
+*   the rights to use, copy, modify, merge, publish, distribute, sublicense,        *
+*   and/or sell copies of the Software, and to permit persons to whom the           *
+*   Software is furnished to do so, subject to the following conditions:            *
+*                                                                                   *
+*   The above copyright notice and this permission notice shall be included in      *
+*   all copies or substantial portions of the Software.                             *
+*                                                                                   *
+*   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR      *
+*   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,        *
+*   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE     *
+*   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER          *
+*   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,   *
+*   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE   *
+*   SOFTWARE.                                                                       *
+*                                                                                   *
+*************************************************************************************/
+
+#ifndef RTTR_LIBRARY_P_H_
+#define RTTR_LIBRARY_P_H_
+
+#include "rttr/detail/base/core_prerequisites.h"
+#include "rttr/string_view.h"
+
+#include "rttr/type.h"
+#include "rttr/property.h"
+#include "rttr/method.h"
+#include "rttr/detail/registration/registration_state_saver.h"
+
+#if RTTR_PLATFORM == RTTR_PLATFORM_WINDOWS
+    #include <windows.h>
+#endif
+
+#include <atomic>
+
+namespace rttr
+{
+namespace detail
+{
+
+/*!
+ * This class contains the private implementation of the \ref library class.
+ *
+ */
+class RTTR_LOCAL library_private
+{
+public:
+    library_private(string_view file_name, string_view version)
+    :   m_file_name(file_name),
+        m_version(version),
+        m_handle(nullptr)
+    {
+    }
+
+    ~library_private() = default;
+
+    bool load()
+    {
+        if (m_handle)
+        {
+            ++m_load_count;
+            return true;
+        }
+        m_state_saver.save_state_begin();
+
+        auto result = load_native();
```

Comment:
if I understand correctly and as far as I see so far: the load method can be called from multiple threads?
if so, isnt there some troubles with thread-safety, maybe I am missing something: A thread A can potentially be in load_native() which is just about to set the m_handle and another thread B can be at line 73 which then loads the library again, right? (thats not what you want right?)

Reply:
You are absolutely right. This is missing. There were some compilers which didn't support threading at all, I think it was minGW 4.9.1.
Mhm, I guess I have to lift the requirements for this. Or at a fallback with using boost. 
https://github.com/meganz/mingw-std-threads
When you know something, please tell me.

Reply:
hm.. ok, I am suspecting that the whole function should be guarded (isnt it?) (not sure about the `m_state_saver`.) So to be absolutely sure that no other thread is currently trying to load the library. 
(I have never loaded a library with dlopen so far, what I understand is that if one thread loads it, the other thread can use it too because its in shared memory, because of your singelton implementation).

I think the only thing we need here is some mutex or thread guard for the `library_manager`
such to prevent race condition of threads loading / unloading the library at the same time (isnt it?). `std::mutex`?